theseus/geometry/utils/utils.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+from typing import Any
+
+import torch
+
+
+class LieGroupContext(object):
+    contexts = threading.local()
+
+    @classmethod
+    def get_context(cls):
+        if not hasattr(cls.contexts, "use_lie_tangent"):
+            cls.contexts.use_lie_tangent = False
+        return cls.contexts.use_lie_tangent
+
+    @classmethod
+    def set_context(cls, use_lie_tangent: bool):
+        cls.contexts.use_lie_tangent = use_lie_tangent
+
+
+class set_lie_tangent_enabled(object):
+    def __init__(self, mode: bool) -> None:
+        self.prev = LieGroupContext.get_context()
+        LieGroupContext.set_context(mode)
+
+    def __enter__(self) -> None:
+        pass
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        LieGroupContext.set_context(self.prev)
+
+
+class enable_lie_tangent(object):
+    def __enter__(self) -> None:
+        self.prev = LieGroupContext.get_context()
+        LieGroupContext.set_context(True)
+
+    def __exit__(self, typ, value, traceback) -> None:
+        LieGroupContext.set_context(self.prev)
+
+
+class no_lie_tangent(LieGroupContext):
+    def __enter__(self):
+        self.prev = super().get_context()
+        LieGroupContext.set_context(False)
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        LieGroupContext.set_context(self.prev)
+
+
+class LieGroupTensor(torch.Tensor):
+    from torch._C import _disabled_torch_function_impl
+
+    __torch_function__ = _disabled_torch_function_impl
+
+    def __new__(cls, group):
+        return torch.Tensor._make_subclass(cls, group.data)
+
+    def __init__(self, group):
+        self.group_cls = type(group)
+
+    def add_(self, update, alpha=1):
+        if LieGroupContext.get_context():
+            group = self.group_cls(data=self.data)
```

Comment:
One concern here and for the functions below is that this hard codes the assumption that a call to `add_` wants to add the gradient, which might not be the case for all of the optimization algorithms in Python. 

Is there a way to change this implementation so that within the context `lie_tensor.grad` value is replaced by its projection to tangent space, and `add_` is just a call to `retract`?

Reply:
That's what I previously tried. However, we might might implement `torch.optim` algorithms in that case. This is something similar to the notion of `virtual function` and `retract()` will be called only if`LieGroupContext.get_context()` is `True`.

Reply:
BTW, note that `LieGroupContext.get_context()` is `False` by default and the users need something like `with theseus.lie_tangent_enabled` to trigger `retract`.

Reply:
Yes, I'm talking about the case where the `LieGroupContext.get_context()` is `True`. Even in that scenario, we don't know if some torch optimizer step will try to do an in-place addition whose argument is not the gradient, but some other vector (e.g. ,a different descent direction), and in this case doing the projection would be wrong, no? Or is this projection correct for any vector in eucildean space, not just the exact gradient?

Reply:
See #79